engines/terraform/resolve.go

```diff
@@ -84,13 +57,111 @@ func (td *TerraformDeployment) resolveValue(intentName string, value interface{}
 	}
 }
 
+func (td *TerraformDeployment) resolveToken(intentName string, specRef *SpecReference) (interface{}, error) {
+	switch specRef.Source {
+	case "infra":
+		if len(specRef.Path) < 2 {
+			return nil, fmt.Errorf("infra token requires at least 2 path components")
+		}
+
+		refName := specRef.Path[0]
+		attribute := strings.Join(specRef.Path[1:], ".")
+
+		infraResource, err := td.resolveInfraResource(refName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve infrastructure resource %s: %w", refName, err)
+		}
+
+		result := infraResource.Get(jsii.String(attribute))
+		if result == nil {
+			return nil, fmt.Errorf("attribute '%s' not found on infrastructure resource '%s' (type: %s)", attribute, refName, *infraResource.Node().Id())
+		}
+		return result, nil
+
+	case "self":
+		if len(specRef.Path) < 1 {
+			return nil, fmt.Errorf("self token requires at least 1 path component")
```

Comment:
```suggestion
			return nil, fmt.Errorf("'%s' references 'self' without specifying a variable. All reference must include a variable name e.g. ${self.my_var}", intentName)
```

Reply:
There may be something better than this still, just a suggestion on how to make the error easier to fix for the user.